### PR TITLE
UHM-4640: Discharge should un-enroll from COVID-19 program

### PR DESCRIPTION
**Pull request by mogoodrich**

### CODE REVIEW

**Comment by mogoodrich** [comment]
Thanks @lnball !

File: api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import org.openmrs.module.emrapi.disposition.Disposition;
+import org.openmrs.module.emrapi.disposition.DispositionDescriptor;
+import org.openmrs.module.emrapi.disposition.DispositionService;
+import org.openmrs.module.htmlformentry.CustomFormSubmissionAction;
+import org.openmrs.module.htmlformentry.FormEntrySession;
+
+import java.util.Date;
+import java.util.List;
+
+public class CloseCovidProgramAction implements CustomFormSubmissionAction {
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java:23]: `ExitFromCovidProgramAction`? Hopefully the hospital keeps the program open :)

File: api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    private static String CONCEPT_DISCHARGE = "TREATMENT STOPPED";
+
+    @Override
+    public void applyAction(FormEntrySession formEntrySession) {
+        closeCovidProgram(formEntrySession, Context.getProgramWorkflowService(), Context.getService(DispositionService.class), Context.getConceptService());
+    }
+
+
+    private void closeCovidProgram(FormEntrySession formEntrySession, ProgramWorkflowService programWorkflowService, DispositionService dispositionService, ConceptService conceptService) {
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java:50]:
```suggestion
    private void closeCovidProgram(FormEntrySession formEntrySession,
                                   ProgramWorkflowService programWorkflowService,
                                   DispositionService dispositionService,
                                   ConceptService conceptService) {
```

File: api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        // Jan 12 is the correct completion date and update
+        // however, if a discharge form is entered on Jan 16, the program from Jan 11 to Jan 14 is *not* updated
+
+        List<PatientProgram> candidates = programWorkflowService.getPatientPrograms(patient, covid, null,
+                encounterDate, encounterDate, null, false);
+
+        if (candidates != null) {
+
+            if (candidates.size() > 1) {
+                log.warn("More than one COVID program enrolment for patient " + patient.getId() + " on date " + encounterDate);
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java:69]:
```suggestion
                log.warn("More than one COVID program enrollment for patient " + patient.getId() +
                         " on date " + encounterDate + ". Now unenrolling from all of them.");
```

File: api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+            for (PatientProgram patientProgram : candidates) {
+
+                // set the date completed to the encounter date
+                patientProgram.setDateCompleted(encounterDate);
+
+                // find the disposition
+                DispositionDescriptor dispositionDescriptor = dispositionService.getDispositionDescriptor();
+
+                Disposition disposition = null;
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java:80]:
```suggestion
                Disposition disposition;
```

File: api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                for (Obs candidate : encounter.getObsAtTopLevel(false)) {
+                    if (dispositionDescriptor.isDisposition(candidate)) {
+                        disposition = dispositionService.getDispositionFromObsGroup(candidate);
+                        break;
+                    }
+                }
+
+                // set program outcome based on dispositoion
+                if (disposition != null) {
+                    if (disposition.getUuid().equals(DISPOSITION_DEATH)) {
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java:91]: I wouldn't expect `disposition.getUuid()` to ever equal `"markPatientDead"`?

**Comment by brandones** [comment]
@mogoodrich A few nits and one major issue

File: api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java (hunk 1 above)
Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java:23]: I can add "Patient", ie "ExitPatientFromCovidProgramAction"

File: api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java (hunk 5 above)
Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java:95]: That's a good expectation, but it does... I don't know why that field is called uuid, it probably should be called key.  I should comment, however...

https://github.com/PIH/openmrs-config-zl/blob/master/configuration/pih/pih-dispositions-haiti.json#L50

File: api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java (hunk 4 above)
Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java:84]: Without setting dispositon to null here you get a "variable might not have been initialized" warning on line 90 (which would happen if it never may any trips through the loop that starts on line 82)

**Comment by mogoodrich** [comment]
Thanks @brandones  I committed 2 of your suggested changes and commented about the other.

I renamed the class to something that makes a little more sense.

And I added a comment about the weird usage of uuid...

File: api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java (hunk 5 above)
Comment (brandones) [reply api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java]: :open_mouth:

**Pull request merged**